masschallenge/impact-api
# AC-4764: History API calls

[AC-4764](https://masschallenge.atlassian.net/browse/AC-4764)

Work done:
- Two new high level API calls were created, /v1/user_history/<id> and /v1/organization_history/<id>.
- These API calls return a chronological history of the user or organization according to the current accelerate database including events as described in the ticket

To test:
- Through the browsable API, logged in a as user with v1_clients permissions, you should be able to run the API calls.
- Interesting examples:
- http://localhost:8000/api/v1/user_history/33445/ is a user that has been a judge and a mentor
- http://localhost:8000/api/v1/user_history/42239/ is a user that has been a finalist
- http://localhost:8000/api/v1/organization_history/35/ is an organization that has only been an entrant
- http://localhost:8000/api/v1/organization_history/6668/ is an organization that has become a finalist
- http://localhost:8000/api/v1/organization_history/16576/ is an organization that is a partner

Something to focus on in the code are the cases where there is a lastest_datetime.  This happens when there is potential ambiguity regarding when an event happened due to potentially missing created_at data.  In these cases we attempt to calculate a reasonable guess of when an event happened.  http://localhost:8000/api/v1/user_history/33445/ is a good example for this since it include events that occurred both before and after we were keeping reliable created_at information.  Note that in the tests to recreate situations where created_at is None, it has to be done explicitly using .save() since it cannot be set through the factory arguments.
